---
title: "Single-cell RNA-squencing ~ Session 2 <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_course_template/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---
```{r setup, include=FALSE}
suppressPackageStartupMessages(require(knitr))
suppressPackageStartupMessages(require(Seurat))
suppressPackageStartupMessages(require(phaetmap))
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(require(bioMart))
suppressPackageStartupMessages(require(SeuratWrappers))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```

---
##  Data analysis with [Seurat](https://satijalab.org/seurat/)

[Seurat](https://satijalab.org/seurat/) is a powerful package for single-cell data analysis. It supports many common analysis of single-cell RNA-Seq and cross connecction with many useful packages through [SeuratWrapper](https://github.com/satijalab/seurat-wrappers). In this session, we will demonstrate our regular workflow for single-cell RNA-seq data analysis with Seurat.

---
## Outline
- Read Cell Ranger results into Seurat object
- Generate QC
- Evaluate cell cycle
- Normalize data and make clustering
- Marker genes for each cluster
- Advanced visualization

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Create Seurat object and Generare QC plots

---
"    
  )
  
}

```

## Cell Ranger counting matrix
- The data is avialable on [BOX]()
```{r loadCR2Seurat,echo=TRUE,eval=FALSE,include=TRUE}
tar_dir <- "path to raw data"
```

---
## Load Cell Ranger results into Seurat object
- Loading data
- Estimate mitochondrial contents
- Subset 1000 cells

```{r loadCR2Seurat2,echo=TRUE,eval=FALSE,include=TRUE}
library(Seurat)
samID <- "CTRL"
X10_file <- Seurat::Read10X(tar_dir)
obj <- CreateSeuratObject(counts = X10_file, project = samID, min.cells = 5,min.features = 200)
obj$dset <- samID
obj <- Seurat::RenameCells(obj,add.cell.id=samID)
#
obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^mt-")
#
cellID <- rownames(obj@meta.data)
cellID_sel <- sample(cellID,1000,replace = FALSE)
obj_sel <- obj[,cellID_sel]
saveRDS(obj_sel,file.path("data",paste0("scSeq_",samID,"_1kCell_ori.rds")))
```
```{r loadCR2Seurat3,echo=TRUE,eval=FALSE,include=FALSE}
#obj <- readRDS("../data/scSeq_CTRL_1kCell_ori.rds")
obj <- readRDS("data/scSeq_CTRL_1kCell_ori.rds")
```

---
## Information in the objects
```{r objInfo,echo=TRUE,eval=TRUE,include=TRUE}
obj
```

---
## Cell information
Containing cell barcode, mitochodiral contents, datasets, and clustering etc
```{r objInfo_cell,echo=TRUE,eval=TRUE,include=TRUE}
head(obj@meta.data)
```

---
## Counting Matrix
- raw count matrix
- scaled or normalized data 
```{r objInfo_count,echo=TRUE,eval=TRUE,include=TRUE}
message("raw counts")
head(obj@assays$RNA@counts)
message("scaled data")
head(obj@assays$RNA@data)
```

---
## Evaluate read counts, genes counts, and mitochondrial contents
- read counts: nCount_RNA
- gene counts: nFeature_RNA
- mitochondrial content: percent.mt
```{r eval_readCount_geneCount_mitoCont,echo=TRUE,eval=TRUE,include=TRUE}
# obj <- readRDS("../data/scSeq_CTRL_1kCell_ori.rds")
VlnPlot(obj,features = c("nCount_RNA","nFeature_RNA","percent.mt"),pt.size = 0.2)
```

---
## read counts vs genes counts
```{r eval_readCount_geneCount,echo=TRUE,eval=TRUE,include=TRUE}
FeatureScatter(obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

---
## read counts vs mitochondrial contents
```{r eval_readCount_mitoCont,echo=TRUE,eval=TRUE,include=TRUE}
FeatureScatter(obj, feature1 = "nCount_RNA", feature2 = "percent.mt")
```

---
## Remove potential cell debris
- high mitochondrial content and low read counts
```{r rmMito,echo=TRUE,eval=TRUE,include=TRUE}
summary(obj@meta.data$percent.mt)
mt_cutH <- 10
obj <- subset(obj,subset = percent.mt < mt_cutH)
```

---
## Remove potential cell debris
```{r rmMito_vln,echo=TRUE,eval=TRUE,include=TRUE}
VlnPlot(obj,features = c("nCount_RNA","nFeature_RNA","percent.mt"),pt.size = 0.2)
```

---
## Remove potential cell debris
```{r rmMito_scatter,echo=TRUE,eval=TRUE,include=TRUE}
FeatureScatter(obj, feature1 = "nCount_RNA", feature2 = "percent.mt") 
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Cell cycle phase determination

---
"    
  )
  
}

```

## Cell cycle
- In Seurat, the default marker genes for S phase and G2M pahse are stored in *cc.gene* list
- This list is based on human data.
- We can convert it into mouse gene symbol by using *bioMart*
```{r cellCycle_conv,echo=TRUE,eval=FALSE,include=TRUE}
convertHumanGeneList <- function(x){
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  genesV2 = getLDS(attributes = c("hgnc_symbol"), 
                   filters = "hgnc_symbol", 
                   values = x , 
                   mart = human,
                   attributesL = c("mgi_symbol"), 
                   martL = mouse, uniqueRows=T)
  humanx <- unique(genesV2[, 2])
  return(humanx)}
```

## Cell cycle ~ calculation
```{r cellCycle_est,echo=TRUE,eval=TRUE,include=TRUE}
s_gene <- convertHumanGeneList(cc.genes$s.genes)
g2m_gene <- convertHumanGeneList(cc.genes$g2m.genes)
obj <- CellCycleScoring(obj, s.features = s_gene,
                              g2m.features = g2m_gene, set.ident = TRUE)
obj@meta.data[1:2,]
```

## Cell cycle ~ phase
```{r cellCycle_plot1,echo=TRUE,eval=TRUE,include=TRUE}
yd_dat <- as.data.frame(table(obj@meta.data$dset,obj@meta.data$Phase))
head(yd_dat)
```

## Cell cycle ~ phase
```{r cellCycle_plot2,echo=TRUE,eval=TRUE,include=TRUE}
library(ggplot2)
ggplot(yd_dat,aes(x=Var1,y=Freq,fill=Var2))+geom_bar(stat="identity",position="stack")+labs(x="",y="Counts",fill="Phase")+theme_classic()
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Regression and clustering

---
"    
  )
  
}

```

## Data normalization
Re-scale and regress data to **percent.mt**, **S.score**, and **G2M.score**
```{r regress,echo=TRUE,eval=TRUE,include=TRUE}
obj <- ScaleData(obj,
                 vars.to.regress = c("percent.mt","S.score","G2M.score","Phase"))
```

---
## Princciple Component Analysis
```{r prcomp,echo=TRUE,eval=TRUE,include=TRUE}
set.seed(1000)
obj <- RunPCA(obj, npcs = 30, verbose = FALSE)
```

---
## How many PCs should we take ~ Elbow Plot
- take the point start turning flat
```{r pcSel,echo=TRUE,eval=TRUE,include=TRUE}
ElbowPlot(obj,ndims=30)
#
numPC <- 15
```

---
## Clustering
```{r clust,echo=TRUE,eval=TRUE,include=TRUE}
set.seed(1000)
maxPC <- numPC
obj <- FindNeighbors(obj, reduction = "pca", dims = 1:maxPC)
obj <- FindClusters(obj, resolution = 0.5)
obj <- RunUMAP(obj, reduction = "pca", dims = 1:maxPC)
obj <- RunTSNE(obj, reduction = "pca", dims = 1:maxPC)
```

---
## UMAP plot
```{r seurat_UMAP,echo=TRUE,eval=TRUE,include=TRUE}
DimPlot(obj,reduction="umap") # defaul it "umap"
```

---
## tSNE plot
```{r seurat_tSNE,echo=TRUE,eval=TRUE,include=TRUE}
DimPlot(obj,reduction = "tsne")
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Identify marker genes for each cluster

---
"    
  )
  
}

```

## Marker genes
```{r markGene,echo=TRUE,eval=TRUE,include=TRUE}
obj <- SetIdent(obj,value = "seurat_clusters")
clust.markers <- FindAllMarkers(obj, 
                                only.pos = TRUE,
                                min.pct = 0.25, 
                                logfc.threshold = 0.25)
head(clust.markers)
```

---
## Select top five marker genes for each cluster
```{r markGene_sub,echo=TRUE,eval=TRUE,include=TRUE}
topG <- clust.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
head(topG)
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Advanced plots

---
"    
  )
  
}

```

## Dimplots ~ adjust point size
```{r dimPlot_ptSize,echo=TRUE,eval=TRUE,include=TRUE}
DimPlot(obj,pt.size = 0.2)
```

---
## Dimplots with labeling clusters
```{r dimPlot_label,echo=TRUE,eval=TRUE,include=TRUE}
DimPlot(obj,pt.size = 0.2,label=TRUE)+NoLegend()
```

---
## Marker Gene expression ~ heatmap
```{r markGene_heatmap,echo=TRUE,eval=TRUE,include=TRUE}
DoHeatmap(obj, features = topG$gene) + NoLegend()
```

---
## Marker Gene Expression ~ FeaturePlot
```{r makGene_featurePlot,echo=TRUE,eval=TRUE,include=TRUE}
gene_marker <- c("Krt1","Pthlh","Krt14","Cenpa","Shh")
FeaturePlot(obj,features = gene_marker,pt.size = 0.2)
```

---
## Marker Gene Expression ~ RidgePlot
```{r makGene_RidgePlott,echo=TRUE,eval=TRUE,include=TRUE}
RidgePlot(obj,features = gene_marker)
```

---
## Cell cycle phases by cluster
```{r cellcycle_cluster,echo=TRUE,eval=TRUE,include=TRUE}
tbl <- table(obj@meta.data$seurat_clusters,obj@meta.data$Phase)
tbl_dat <- as.data.frame(tbl)
to <- rowSums(tbl)
names(to) <- rownames(tbl)
tbl_dat$to <- to[match(names(to),tbl_dat$Var1)]
tbl_dat$prop <- tbl_dat$Freq / tbl_dat$to
tbl_dat[1:2,]
```

---
## Cell cycle phases by cluster ~ plot
```{r cellcycle_cluster_plot1,echo=TRUE,eval=FALSE,include=TRUE}
ggplot(tbl_dat,aes(x=Var1,y=prop,fill=Var2))+
  geom_bar(stat="identity",position="stack")+
  labs(x="Seurat_clusters",y="Proportion",fill="Phase")+
  theme_classic()
```

---
## Cell cycle phases by cluster ~ plot
- may need to remove cluster 2
```{r cellcycle_cluster_plot1,echo=FALSE,eval=TRUE,include=TRUE}
ggplot(tbl_dat,aes(x=Var1,y=prop,fill=Var2))+
  geom_bar(stat="identity",position="stack")+
  labs(x="Seurat_clusters",y="Proportion",fill="Phase")+
  theme_classic()
```

---
## Save file
```{r saveRDS_surat,echo=TRUE,eval=TRUE,include=TRUE}
cellID <- rownames(obj@meta.data[!obj@meta.data$seurat_clusters==2])
obj_sub <- obj[,cellID]
saveRDS(obj_sub,"scSeq_Seurat_clean.rds")
```

---
## Convert into SingleCellExperiment
```{r convetSCE,echo=TRUE,eval=TRUE,include=TRUE}
library(SeuratWrappers)
sce <- toSingleCellExperiment(obj_sub)
scce
```

---
## Exercise time
- [Exerise]()
- [Answer]()