---
title: "Single-cell RNA sequencing ~ Session 3<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/scRNA-seq/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---
```{r setup, include=FALSE}
suppressPackageStartupMessages(require(knitr))
suppressPackageStartupMessages(require(DropletUtils))
suppressPackageStartupMessages(require(SingleCellExperiment))
suppressPackageStartupMessages(require(scran))
suppressPackageStartupMessages(require(scater))
suppressPackageStartupMessages(require(scuttle))
suppressPackageStartupMessages(require(scDblFinder))
knitr::opts_chunk$set(echo = TRUE, tidy = T)

```


## Overview

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Loading data and empty droplets

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("#  Loading data and empty drops

---
"    
  )
  
}

```

## Install python libraries
```{r installPythonAndPkgs}

if(!require(reticulate)){
  install.packages("reticulate")
  require(reticulate)
}
if(!dir.exists(reticulate::miniconda_path())){
  install_miniconda()
}
conda_install(
  packages=c("python-igraph","scanpy","louvain")
)



```

---
## Use conda python
```{r usePython}
use_condaenv()
py_config()

```

---
## Load python library
```{r importPython}
sc <- import("scanpy")
sc
```

---
## Install anndata package
```{r installAnnData}
if(!require(anndata)){
  install.packages("anndata")
  require(anndata)
}

```

---
## Download PBMC
```{r downloadPBMC}
pathToDownload <- "http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz"
download.file(pathToDownload,"pbmc3k_filtered_gene_bc_matrices.tar.gz")
utils::untar("pbmc3k_filtered_gene_bc_matrices.tar.gz")
```

---
## Read PBMC
```{r readPBMC}
adata <- sc$read_10x_mtx("filtered_gene_bc_matrices/hg19/",
                         var_names="gene_ids") 

adata
```

---
## Dimensions PBMC
```{r dimensions}
dim(adata)
```

---
## Dimnames - PBMC
```{r dimnames}
colnames(adata)[1:3]
rownames(adata)[1:3]
```

---
## Counts- PBMC
```{r countsMatrix}
countsMatrix <- adata$X
countsMatrix[1:2,1:2]
```

---
## Gene information - PBMC
```{r vardf}
var_df <- adata$var
var_df[1:2,,drop=FALSE]
```

---
## Cell information - PBMC
```{r obsdf}
obs_df <- adata$obs
obs_df
rownames(obs_df)[1:2]
```

---
## Cell information - PBMC
```{r addsomeQC}
sc$pp$filter_cells(adata, min_genes=0)
sc$pp$filter_genes(adata, min_cells=0)
```

---
## Cell information - PBMC
```{r vardf2}
var_df <- adata$var
var_df[1:2,]
```

---
## Cell information - PBMC
```{r obsdf2}
obs_df <- adata$obs
obs_df[1:2,,drop=FALSE]
```

---
## Cell information - PBMC
```{r vardfRe}
var_df$mito <- grepl("MT-",var_df$gene_symbols)
var_df[var_df$mito == TRUE,]
```

---
## Cell information - PBMC
```{r vardfRe2}
adata$var <- var_df
adata$var[1:2,]
```

---
## Cell information - PBMC
```{r santize}
sc$utils$sanitize_anndata(adata)
```

---
## Cell information - PBMC
```{r mitoQC,tidy=FALSE}
sc$pp$calculate_qc_metrics(adata,
                           qc_vars = list("mito"),
                           inplace = TRUE)
adata
```

---
## Cell information - PBMC
```{r moreQC,tidy=FALSE}
adata$var[5:10,]
```

---
## Cell information - PBMC
```{r moreQC2,tidy=FALSE}
adata$obs[1:2,]
```


---
## Cell information - PBMC
```{r plotHiExpr,eval=FALSE}
sc$pl$highest_expr_genes(adata,gene_symbols = "gene_symbols")
```

---
## Cell information - PBMC
```{r plotQCpy2,eval=FALSE}
sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mito'],
             jitter=0.4, multi_panel=True)
```

---
## Cell information - PBMC
```{r plotQCpy3,eval=FALSE}
sc.pl.scatter(adata, x='total_counts', y='n_genes_by_counts')

```



---
## Cell information - PBMC
```{r filtercells}
adata = adata[adata$obs$n_genes < 2500]
adata = adata[adata$obs$n_genes > 200]
adata = adata[adata$obs$pct_counts_mito < 5]
adata
```

---
## Cell information - PBMC
```{r filtergenes}
sc$pp$filter_genes(adata,
                   min_cells=3)
adata
```

---
## Cell information - PBMC
```{r normCells}
sc$pp$normalize_per_cell(adata, counts_per_cell_after=10000)
```

---
## Cell information - PBMC
```{r logCells}
# log transform the data.
sc$pp$log1p(adata)

```

---
## Cell information - PBMC
```{r variableGenes}
# identify highly variable genes.
sc$pp$highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)
# sc$pl$highly_variable_genes(adata)
```


---
## Cell information - PBMC
```{r filterToVariable}
# keep only highly variable genes:
adata = adata$copy()
adata = adata[,adata$var$highly_variable]
adata = adata$copy()
# regress out total counts per cell and the percentage of mitochondrial genes expressed
# adata.copy = adata
# sc$pp$regress_out(adata, list('n_counts', 'pct_counts_mito'),copy=FALSE) #, n_jobs=args.threads)
```

---
## Cell information - PBMC
```{r regress}
sc$pp$regress_out(adata, list('n_counts', 'pct_counts_mito'))

# adata = adata2
# scale each gene to unit variance, clip values exceeding SD 10.
```


---
## Cell information - PBMC
```{r scale}
sc$pp$scale(adata, max_value=10)
```


---
## Cell information - PBMC
```{r pca}
sc$tl$pca(adata, svd_solver='arpack')
```


---
## Cell information - PBMC
```{r neighbours}
sc$pp$neighbors(adata, n_neighbors=15L, n_pcs=50L)
sc$tl$umap(adata)
```

---
## Cell information - PBMC
```{r louvain}
sc$tl$louvain(adata,resolution=0.8)
# 
```

---
## Cell
```{r plotumapCells,eval=FALSE}
sc$pl$umap(adata, color=list('louvain'))
```

---
## Time for an exercise!

Exercise on scRNAseq analysis with Bioconductor can be found [here](../../exercises/exercises/exercise4_exercise.html).

Answers can be found [here](../../exercises/answers/exercise4_answers.html).
