---
title: "Single-cell RNA sequencing ~ Session 3<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/scRNA-seq/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---
```{r setup, include=FALSE}
suppressPackageStartupMessages(require(knitr))
suppressPackageStartupMessages(require(DropletUtils))
suppressPackageStartupMessages(require(SingleCellExperiment))
suppressPackageStartupMessages(require(scran))
suppressPackageStartupMessages(require(scater))
suppressPackageStartupMessages(require(scuttle))
suppressPackageStartupMessages(require(scDblFinder))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```


## Outline
There are more and more Bioconductor packages supporting single-cell data analysis. R Amezquita, A Lun, S Hicks, and R Gottardo wrote an integrated workflow, [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/release/OSCA/), for single-cell data analysis and quality assessment. In this session, we will go through several important QC metrics which can't be made with Seurat. 

- How to differentiate empty droplets?
- How to estimate ambient RNA and remove them?
- How to identify doublets?
- Checking for confounded effect?

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Loading data and empty droplets

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("#  Loading data and empty drops

---
"    
  )
  
}

```

## Load data from Cell Ranger Result
We can load the data in from 10X with the *DropletUtils* function, *read10xCounts()*.
```{r loadSCE,eval=FALSE,include=TRUE,echo=TRUE}
library(DropletUtils)
library(DropletTestFiles)
fname <- "~/filepath/toCellRanger/results"
sce <- read10xCounts(fname, col.names=TRUE)
cellID <- colData(sce)$Barcode

```

---
## Subset dataset
We have a small subset version of this dataset that you can load in from the *data/* directory to try this out. 

```{r loadSCE_io,include=TRUE,echo=FALSE,eval=TRUE}
sce <- readRDS("data/scSeq_CTRL_sceSub.rds")
```

---
## SingleCellExperiment Object

*read10xCounts()* reads the data in as a specialist object called a SingleCellExperiment. 
```{r loadSCE_pres,include=TRUE,echo=TRUE,eval=TRUE}
sce
```

---
## SingleCellExperiment Object
The *colData()* and *rowData()* functions can be used to access experiment metadata. 

```{r loadSCE_pres2,include=TRUE,echo=TRUE,eval=TRUE}
# cell information
colData(sce)[1:2,]
# gene information
rowData(sce)[1:2,]
```

---
## Access UMI counts in each droplet
```{r rankUMI,eval=TRUE,}
bcrank <- barcodeRanks(counts(sce))
bcrank[1:2,]
```

---
## Knee plot
Knee plot is a useful QC plot for single-cell seuqnecing. It reflects the threshold to validate cell for analysis. 
- x-axis is the cell barcodes (droplet) ranked by their UMI counts
- y-axis is the UMI counts in each droplet
- inflection point: the point while UMI counts  start decreasing rapidly
- knee point: the cut-off of UMI counts to differentiate cells valid for analysis 

---
## Knee plot
```{r rankUMI_knee,eval=FALSE,echo=TRUE,include=TRUE}
uniq <- !duplicated(bcrank$rank)
#
plot(bcrank$rank[uniq], bcrank$total[uniq], log="xy",
    xlab="Rank", ylab="Total UMI count", cex.lab=1.2)

abline(h=metadata(bcrank)$inflection, col="darkgreen", lty=2)
abline(h=metadata(bcrank)$knee, col="dodgerblue", lty=2)

legend("bottomleft", legend=c("Inflection", "Knee"), 
        col=c("darkgreen", "dodgerblue"), lty=2, cex=1.2)
```

---
## Knee Plot
```{r rankUMI_knee2,eval=TRUE,echo=FALSE,include=TRUE}
uniq <- !duplicated(bcrank$rank)
#
plot(bcrank$rank[uniq], bcrank$total[uniq], log="xy",
    xlab="Rank", ylab="Total UMI count", cex.lab=1.2)

abline(h=metadata(bcrank)$inflection, col="darkgreen", lty=2)
abline(h=metadata(bcrank)$knee, col="dodgerblue", lty=2)

legend("bottomleft", legend=c("Inflection", "Knee"), 
        col=c("darkgreen", "dodgerblue"), lty=2, cex=1.2)
```

---
## Identify non-empty droplets
  The *emptydrops* function can be used to identify good/bad droplets. 2 keys arguments are:
- limit: lowest counts per cell
- test.ambient: Could be used to estimate ambient RNA contamination

```{r estDroplet,eval=TRUE,echo=TRUE,include=TRUE}
set.seed(100)
limit <- 100   
e.out <- emptyDrops(counts(sce),lower=limit, test.ambient=TRUE)
#
e.out
```

---
## Identify non-empty droplets
```{r estDroplet2,eval=TRUE,echo=TRUE,include=TRUE}
# Testeed by FDR
summary(e.out$FDR <= 0.001)
# Concordance by testing with FDR and limited
table(Sig=e.out$FDR <= 0.001, Limited=e.out$Limited)
```

---
## Distribution of non-empty reads
We can plot the distribution of significance of non-empty reads

```{r estDroplet3,eval=TRUE,echo=TRUE,include=TRUE}
hist(e.out$PValue[e.out$Total <= limit & e.out$Total > 0],
    xlab="P-value", main="", col="grey80") 
```

---
## Subset non-empty droplets
We can filter to just our non-empty droplets using a simple which query on the FDR from *emptyDrops()*. Here we are using a 0.001 cut-off. 
```{r estDroplet4,eval=TRUE,echo=TRUE,include=TRUE}
sce2 <- sce[,which(e.out$FDR <= 0.001)]
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Normalization and clustering

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Normalization and clustering

---
"    
  )
  
}

```

## Counts normalization
```{r countNorm,eval=TRUE,echo=TRUE,include=TRUE}
library(scran)
library(scuttle)
library(scater)
clusters <- quickCluster(sce2)
sce2 <- computeSumFactors(sce2, cluster=clusters)
sce2 <- logNormCounts(sce2)
sce2
```

---
## Identify variable features
```{r featureIdent,eval=TRUE,echo=TRUE,include=TRUE}
set.seed(1000)
# modeling variables
dec.pbmc <- modelGeneVarByPoisson(sce2)
# calcualte top features
top.pbmc <- getTopHVGs(dec.pbmc, prop=0.1)
```

---
## Make TSNE and UMAP plots
```{r plotUMAP,eval=TRUE,echo=TRUE,include=TRUE}
set.seed(1000)
# Evaluate PCs
sce2 <- denoisePCA(sce2, subset.row=top.pbmc, technical=dec.pbmc)
# make TSNE plot
sce2 <- runTSNE(sce2, dimred="PCA")
# make UMAP plot
sce2 <- runUMAP(sce2, dimred="PCA")
```

---
## Graphic based clustering
```{r clust,eval=TRUE,echo=TRUE,include=TRUE}
g <- buildSNNGraph(sce2, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership
colLabels(sce2) <- factor(clust)
#
colData(sce2)
```

---
## UMAP plot
```{r plotUMAP_2,eval=TRUE,echo=TRUE,include=TRUE}
plotUMAP(sce2,colour_by="label")
```

---
## tSNE plot
```{r plotTSNE,eval=TRUE,echo=TRUE,include=TRUE}
plotTSNE(sce2,colour_by ="label")
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Removing Ambient RNA

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Removing Ambient RNA

---
"    
  )
  
}

```

## Ambient RNA
- Cell-free RNAs can contaminate droplets.
- They can be estimated by empty droplets.
```{r estAmb1,eval=TRUE,echo=TRUE,include=TRUE}
# extrat potential ambient RNA and thee estimated score
amb <- metadata(e.out)$ambient[,1]
head(amb)
```

---
## Remove ambient RNA
```{r estAmb2,eval=TRUE,echo=TRUE,include=TRUE}
library(scater)
stripped <- sce2[names(amb),]
out <- removeAmbience(counts(stripped), ambient=amb,groups = colLabels(stripped))
```

---
## Integrate corrected counts
```{r recalAmb,eval=TRUE,echo=TRUE,include=TRUE}
counts(stripped, withDimnames=FALSE) <- out
stripped <- logNormCounts(stripped)
```

---
## Before/After removal
- Hemoglobin A1 (Hba-a1) as example
- In most cases the Hbs are contaminated from residual RBCs
```{r compRMAmb,eval=FALSE,echo=TRUE,include=TRUE}
ensmbl_id <- rowData(sce2)$ID[rowData(sce2)$Symbol=="Hba-a1"]
plotExpression(sce2, x="label", colour_by="label", features=ensmbl_id) +
        ggtitle("Before")

plotExpression(stripped, x="label", colour_by="label", features=ensmbl_id) + 
        ggtitle("After")
```

---
## Before/After removal
Hemoglobin A1 (Hba-a1) as example

.pull-left[
```{r compRMAmb_pres1,eval=TRUE,echo=FALSE,include=TRUE}
ensmbl_id <- rowData(sce2)$ID[rowData(sce2)$Symbol=="Hba-a1"]
plotExpression(sce2, x="label", colour_by="label", features=ensmbl_id) + 
        ggtitle("Before")
```
  ]

.pull-right[
```{r compRMAmb_pres2,eval=TRUE,echo=FALSE,include=TRUE}
plotExpression(stripped, x="label", colour_by="label", features=ensmbl_id) + 
        ggtitle("After")
```
  ]
  
---
## Before/After removal
- Krt17 as example
```{r compRMAmb2,eval=FALSE,echo=TRUE,include=TRUE}
ensmbl_id <- rowData(sce2)$ID[rowData(sce2)$Symbol=="Krt17"]
plotExpression(sce2, x="label", colour_by="label", features=ensmbl_id) +  
        ggtitle("Before")
plotExpression(stripped, x="label", colour_by="label", features=ensmbl_id) + 
        ggtitle("After")
```

---
## Before/After removal
Krt17 as example

.pull-left[
```{r compRMAmb2_pres,eval=TRUE,echo=FALSE,include=TRUE}
ensmbl_id <- rowData(sce2)$ID[rowData(sce2)$Symbol=="Krt17"]
plotExpression(sce2, x="label", colour_by="label", features=ensmbl_id) + 
        ggtitle("Before")
```
  ]

.pull-right[
```{r compRMAmb2_pres2,eval=TRUE,echo=FALSE,include=TRUE}
plotExpression(stripped, x="label", colour_by="label", features=ensmbl_id) + 
        ggtitle("After")
```
  ]

---
## Normalization and clustering
```{r normClust,eval=FALSE,echo=TRUE,include=TRUE}
dec <- modelGeneVar(stripped)
hvgs <- getTopHVGs(dec,n=1000)
stripped <- runPCA(stripped, ncomponents=10, subset_row=hvgs)
stripped <- runUMAP(stripped, dimred="PCA")
g <- buildSNNGraph(stripped, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership
colLabels(stripped) <- factor(clust)
plotUMAP(stripped,colour_by="label")
```

---
## Normalization and clustering
```{r normClust2,eval=TRUE,echo=FALSE,include=TRUE,fig.height=4,fig.width=4}
dec <- modelGeneVar(stripped)
hvgs <- getTopHVGs(dec,n=1000)
stripped <- runPCA(stripped, ncomponents=10, subset_row=hvgs)
stripped <- runUMAP(stripped, dimred="PCA")
g <- buildSNNGraph(stripped, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership
colLabels(stripped) <- factor(clust)
plotUMAP(stripped,colour_by="label")
```

---
## Save result
Again we can save our results for later using a Rds object.
```{r rmAmb_store,eval=TRUE,echo=TRUE,include=TRUE}
saveRDS(stripped,"data/scSeq_CTRL_sceSub_rmAmbRNA.rds")
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Removing Doublets

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Removing Doublets

---
"    
  )
  
}

```


## Doublets
- Doublets means two or more cells clumped in a single droplet. Thus, the read counts and genes detected in this droplet would be much higher than other droplets.
- Here, we demonstrate how to identify doublets using simulation with [scran](https://bioconductor.org/packages/release/bioc/manuals/scran/man/scran.pdf)

---
## Estimate doublets
The *computeDoubletDensity()* fucntion can be used on your SingleCellExperiment object to estimate doublets. 
```{r estDoublet,eval=TRUE,echo=TRUE,include=TRUE}
dbl.dens <- computeDoubletDensity(stripped, #subset.row=top.mam, 
    d=ncol(reducedDim(stripped)),subset.row=hvgs)
summary(dbl.dens)
stripped$DoubletScore <- dbl.dens
```

---
## Plot doublets scores ~ UMAP
We can project the doublet scores onto our UMPA to see if doublet correlates with the data distribution.
```{r doublerScore,eval=TRUE,echo=TRUE,include=TRUE}
plotUMAP(stripped,colour_by="DoubletScore")
```

---
## Plot doublets scores by cluster
Likewise we can look for over representation of doublets in clusters.
```{r doublerScorebyClust,eval=TRUE,echo=TRUE,include=TRUE,fig.height=3.5,fig.width=5}
plotColData(stripped, x="label", y="DoubletScore", colour_by="label")+
  geom_hline(yintercept = quantile(colData(stripped)$DoubletScore,0.95),lty="dashed",color="red")
```
- No clusters have significantly higher doublet scores than other clusters. No clusters would be removed.
- Red dash line represented 95% quantile of doublet score. The cells with higher doublet score than this cut-off would be removed.

---
## Remove doublets
We can clean our data up based on this 95% quantile cut-off. 
```{r rmDoublet,eval=TRUE,echo=TRUE,include=TRUE}
cut_off <- quantile(stripped$DoubletScore,0.95)
stripped$isDoublet <- c("no","yes")[factor(as.integer(stripped$DoubletScore>=cut_off),levels=c(0,1))]
table(stripped$isDoublet)
sce_clean <- stripped[stripped$isDoublet=="no",]
saveRDS(sce_clean,"data/scSeq_CTRL_sceSub_rmAmbRNA_rmDoublet.rds")
```



---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# QC plots after clearance

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# QC plots after clearance

---
"    
  )
  
}

```


## QC plots
- Mitochondrial content
- Genes detected
- Reads per cell
```{r evaQC_cal,eval=TRUE,echo=TRUE,include=TRUE}
library(scater)
mtGene <- rowData(sce_clean)$ID[grepl(rowData(sce_clean)$Symbol,pattern = "mt-")]
is.mito <- names(sce_clean) %in% mtGene
sce_clean <- addPerCellQC(sce_clean, subsets=list(Mito=is.mito))
```

---
## QC plots - Read counts


Post-clean
```{r qc_mrgr,eval=TRUE,echo=TRUE,include=TRUE,fig.height=4,fig.width=4}
plotColData(sce_clean,x="label", y="sum", colour_by="label")+ggtitle("read counts")
```


---
## QC plots - Gene counts

Post-clean
```{r qc_mrg2r,eval=TRUE,echo=TRUE,include=TRUE,fig.height=4,fig.width=4}
plotColData(sce_clean,x="label", y="detected", colour_by="label")+ggtitle("gene counts")
```


---
## QC plots - Mitochondrial %

Post-clean
```{r qc_mrg3r,eval=TRUE,echo=TRUE,include=TRUE,fig.height=4,fig.width=4}
plotColData(sce_clean,x="label", y="subsets_Mito_percent", colour_by="label")+ggtitle("mitocondrial content")
```


---
## QC plots ~ comparison
- Mitochondrial contents vs read counts
```{r qc_complex,eval=TRUE,echo=TRUE,include=TRUE,fig.height=4,fig.width=4}
plotColData(sce_clean,x="sum",y="subsets_Mito_percent",colour_by="label")+ggtitle("is.mito vs read counts")
```

---
## QC plots ~ comparison
- Gene counts vs read counts
```{r qc_complex2,eval=TRUE,echo=TRUE,include=TRUE,fig.height=4,fig.width=4}
plotColData(sce_clean,x="sum",y="detected",colour_by="label")+ggtitle("gene counts vs read counts")
```

---
## Estimate variance explaination
- Clustering (label), mitochondrial content (subsets_Mito_percent), doublets (DoubletScore), read counts (sum), and gene counts (detected) were tested.
- We would suppose "label" (clustering) would explain more variances than other controls.
```{r varExp,eval=T,echo=TRUE,include=TRUE}
vars <- getVarianceExplained(sce_clean, 
    variables=c("DoubletScore","label","sum","detected","subsets_Mito_percent"))
```
```{r varExp2,eval=F,echo=TRUE,include=TRUE}
plotExplanatoryVariables(vars)
```


---
## Estimate variance explaination

```{r varExp3,eval=TRUE,echo=TRUE,include=TRUE,fig.height=4,fig.width=4}
plotExplanatoryVariables(vars)
```


---
## Time for an exercise!

Exercise on scRNAseq analysis with Bioconductor can be found [here](../../exercises/exercises/exercise2_exercise.html).

Answers can be found [here](../../exercises/answers/exercise2_answers.html).
