---
title: "Single-cell RNA-squencing ~ Session 1<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_course_template/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```

---
## Overview
In this course we are goining to introduce basic analysis for single-cell RNA-Seq, especially in 10X system. The course is divided into three sessions. In the first session, we will introduce how to interpret the Cell  Ranger QC report and make analysis with LOUPE  Browser. We will also demonstrate the customized analysis we can support right know. In the Second session, we will demonstrate how to process scRNA-Seq data and make QC reports with Seurat. In the third session, we will discuss how to make advanced analysis with Bioconductor packages. 

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Cell ranger results

---
"    
  )
  
}

```

## Cell Ranger
[Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) is widely used software for single-cell sequencing data analysis. It supports various kinds of analysis for 10X single-cell sequencing data.

- Alignment and counting
- Analyse VDJ changes
- Aggregate datasets
- Convert bcl files (raw files) into fastq files
- Generate customized GTF and Genome


---
## Basic reports of Cell Ranger
- [QC Report]("data/WT.html")
- cloupe file: can be processed by [LOUPE Browser](https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser)

---

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("Customized analysis

---
"    
  )
  
}

```

## Dmonstration
Here we use an article published by Fuchs Lab as an example to deomostrate the customized analysis we have done. Please access this [article](https://elifesciences.org/articles/56980). 
- General QC plots
- Data integration and normalization
- Clustering
- Annotate cell types
- Pseudotime analysis
- Differential gene expression and single-cell pathway analysis

---
## Fhe next two sessions
We will also use the data of the eLife paper in next two sessions. You can download the data from GEO [link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE147415) or from BOX [link](https://rockefeller.box.com/s/hdy0vex38dlph5pglafrkw68b7uemndr)
