---
title: "Single-cell RNA sequencing ~ Session 2<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/scRNA-seq/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
editor_options: 
  chunk_output_type: console
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```

## Overview

In this course we are going to introduce basic analysis for single-cell RNAseq, with a specific focus on the *10X system*. The course is divided into multiple sessions. 

In this second session, we will introduce Seurat and Bioconductor methodologies for basic scRNA-seq QC, filtering and initial analysis of single samples.
---
## The data

For these sessions we are going to make use of two datasets.

The first set will be from the recent paper [**Enteroendocrine cell lineages that differentially control feeding and gut motility**](https://elifesciences.org/articles/78512).  
This contains scRNA data from either Neurod1 and Neurog3 expressing enteroendocrine cells.

The second dataset is the classic example from PBMC cells.

---
## Seurat and Bioconductor

The Seurat R package and associated packages offer an R based and well established methodology for the analysis of single cell RNA-seq, ATAC-seq and many other single-cell sequencing methodologies.

[Seurat page](https://satijalab.org/seurat/)

Bioconductor has a set of interrelated and highly connected software packages for single cell RNAseq as well as integrating with the rest of the Bioconductor software ecosystem.
[OSCA book](https://bioconductor.org/books/3.18/OSCA/)


In this session we will run through basic analyses using both of the methodologies to show both their equvalence in some parts of the analysis as well as their distinct functionalities in others.

In practice, you may select methods from either softwares and convert between the two as required.

---
# The Example Data

For this session we will need the filtered and raw expression data for the Neurod1 datasets.

The filtered matrix can be found [here](https://rubioinformatics.s3.amazonaws.com/scRNA_graduate/SRR_NeuroD1/outs/NeuroD1_filtered_feature_bc_matrix.h5)

The raw matrix can be found here [here](https://rubioinformatics.s3.amazonaws.com/scRNA_graduate/SRR_NeuroD1/outs/raw_feature_bc_matrix.h5)

---

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Bioconductor methods

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Bioconductor methods

---
"    
  )
  
}

```

---
# Data import (Bioconductor) 

First we need to load the DropletUtils package to read and handle our Droplet 10X data.

```{r,fig.width=7,fig.height=4}

library(DropletUtils)

```

---
# Read10x

We can then read in the filtered matrix containing data on droplets marked as cells by CellRanger.

We set the row.names argument for easier interpretation.

```{r,fig.width=7,fig.height=4}

h5file <- "https://rubioinformatics.s3.amazonaws.com/scRNA_graduate/SRR_NeuroD1/outs/filtered_feature_bc_matrix.h5"
local_h5file <- basename(h5file)
download.file(h5file,local_h5file)

sce.NeuroD1_filtered <- read10xCounts(local_h5file, col.names=TRUE,row.names = "symbol")

class(sce.NeuroD1_filtered)

```

```{r,include=FALSE}
unlink(h5file)
```

---

## SingleCellExperiment Object
The created SingleCellExperiment Object is much like our SummarizedExperiment object

```{r,fig.width=7,fig.height=4}
sce.NeuroD1_filtered
```

---

## SingleCellExperiment Object
The *colnames()* and *rownames()* functions can be used to access column names (cell-barcodes) and row names (gene identifiers)

```{r loadSCE_pres222,include=TRUE,echo=TRUE,eval=TRUE}
# cell information
colnames(sce.NeuroD1_filtered)[1:2]
# gene information
rownames(sce.NeuroD1_filtered)[1:2]
```

---
## SingleCellExperiment Object
The *colData()* and *rowData()* functions can be used to access experiment metadata. 

```{r loadSCE_pres12,include=TRUE,echo=TRUE,eval=TRUE}
# cell information
colData(sce.NeuroD1_filtered)[1:2,]
# gene information
rowData(sce.NeuroD1_filtered)[1:2,]
```

---

## SingleCellExperiment Object

A reducedDim and reducedDimNames slots at present remains unfilled.

The metadata slot contains the Sample names. Here all these cells came from single sample

```{r loadSCE_pres2,include=TRUE,echo=TRUE,eval=TRUE}
# cell information
reducedDimNames(sce.NeuroD1_filtered)
# gene information
metadata(sce.NeuroD1_filtered)
```


---
## The knee plot

One of the first plots we may want to recreate from the QC is the knee plot.

To do this we will need the unfiltered matrix too. 

The unfiltered SingleCellExperiment contains over a million droplets.


```{r loadSCE_pres3,include=TRUE,echo=TRUE,eval=TRUE}
h5file <- "https://rubioinformatics.s3.amazonaws.com/scRNA_graduate/SRR_NeuroD1/outs/raw_feature_bc_matrix.h5"
local_h5file <-"NeuroD1_raw_feature_bc_matrix.h5"
download.file(h5file,local_h5file)
sce.NeuroD1_unfiltered <- read10xCounts(local_h5file, col.names=TRUE)
sce.NeuroD1_unfiltered
```

```{r,include=FALSE,eval=FALSE}
unlink(local_h5file)
```

---
## Barcode ranks

We can use the *barcodeRanks* function to create a table of barcode ranks and total reads per droplet.


```{r,fig.width=7,fig.height=4}
bcrank <- barcodeRanks(counts(sce.NeuroD1_unfiltered))
bcrank
```

---
## Retrieve filtered barcodes

We then filter for droplets with the same rank for visualisation purposes.

```{r,fig.width=7,fig.height=4}
bcrank$filtered <- rownames(bcrank) %in% colnames(sce.NeuroD1_filtered)
bc_plot <- as.data.frame(bcrank)
bc_plot <- bc_plot[order(bc_plot$filtered,decreasing=TRUE),]
bc_plot <- bc_plot[!duplicated(bc_plot$rank),]
```

---
## Plotting barcode rank

We can now recreate the knee plot from CellRanger.

```{r,fig.width=7,fig.height=4}
require(ggplot2)
  ggplot(bc_plot,aes(x=rank,y=total,colour=filtered,alpha=0.001))+
  geom_point()+
  scale_y_log10()+
  scale_x_log10()+
  theme_minimal()+
  geom_hline(yintercept = metadata(bcrank)$inflection,colour="darkgreen",linetype=2)+
  geom_hline(yintercept = metadata(bcrank)$knee,colour="dodgerblue",linetype=2)

```

---
## Filtering droplets using Droplet utils

Although CellRanger provides a filtered Droplet matrix, you may wish to provide your own cut-off based on knee plots generated and/or use the emptyDrops method in Droplet utils to identify cell containing droplets.

```{r,eval=FALSE,echo=TRUE}
e.out <- emptyDrops(counts(sce.NeuroD1_unfiltered))
table(e.out)
head(e.out)
```

```{r,eval=TRUE,echo=FALSE}
e.out <- readRDS(system.file("extdata/data/e.out.RData", mustWork = TRUE, package = "scRNASeq"))
head(e.out)
```

```{r,include=FALSE,eval=FALSE}
unlink(local_h5file)
```

---

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Bioconductor  -  QC

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Bioconductor  -  QC

---
"    
  )
  
}

```

## Working with CellRanger filtered matrix

From this point we will use the Droplets filtered using Cellranger

```{r,fig.width=7,fig.height=4}
sce.NeuroD1_filtered
```
  
---
## Add QC metrics

To assess QC we first want to identify mitochondrial and ribosomal genes.

High mitochondrial gene expression within a cell is often used as a marker of dying cells.

```{r,fig.width=7,fig.height=4}
is.mito <- grepl("^MT",rowData(sce.NeuroD1_filtered)$Symbol)
# is.ribo <- grepl("^Rps",rowData(sce.NeuroD1_filtered)$Symbol)

table(is.mito)
# table(is.ribo)

```
  
---
## Add QC metrics

The scuttle package has functions for normalisation, transformation and QC.

Here we us the addPerCellQCMetrics function to update our SingleCellExperiment object with QC information.

```{r,fig.width=7,fig.height=4}
library(scuttle)
sce.NeuroD1_filtered <- addPerCellQCMetrics(sce.NeuroD1_filtered, 
                                            subsets=list(Mito=is.mito))
sce.NeuroD1_filtered
```

---
## QC in colData

The QC data per Droplet has been added to additional columns in the colData slot of our SingleCellExperiment.

```{r,fig.width=7,fig.height=4}
qc_df <- colData(sce.NeuroD1_filtered)
head(qc_df,n=2)
```


---
## Plotting QC data

We can then plot this using the plotColData function from the Scater package.

First we plot the distritbution of total reads per cell.

```{r,fig.width=7,fig.height=4}
library(scater)
plotColData(sce.NeuroD1_filtered,x="Sample", y="sum")
```

---
## Plotting QC data

Then we plot the distritbution of total detected genes per cell.

```{r,fig.width=7,fig.height=4}
library(scater)
plotColData(sce.NeuroD1_filtered,x="Sample",y="detected")

```

---
## Plotting QC data

Then we plot the distritbution of percent of mitochondrial reads per cell.

```{r,fig.width=7,fig.height=4}
library(scater)
plotColData(sce.NeuroD1_filtered,x="Sample",y="subsets_Mito_percent")
```

---
## Plotting QC data

It also can be useful to capture the QC columns against each other as scatterplots

```{r,fig.width=7,fig.height=4}
library(scater)
p1 <- plotColData(sce.NeuroD1_filtered,x="sum",y="detected")
p2 <- plotColData(sce.NeuroD1_filtered,x="sum",y="subsets_Mito_percent")
p3 <- plotColData(sce.NeuroD1_filtered,x="detected",y="subsets_Mito_percent")
gridExtra::grid.arrange(p1,p2,p3,ncol=3)
```

---
## Removing low quality data

At this point we may want to remove low quality data.

Typically droplet filters on mitochondrial read content, total detect genes and total read counts are applied.

```{r,fig.width=7,fig.height=4}
qc.high_lib_size <- colData(sce.NeuroD1_filtered)$sum > 125000
qc.min_detected <- colData(sce.NeuroD1_filtered)$detected < 200
qc.mito <- colData(sce.NeuroD1_filtered)$subsets_Mito_percent > 25
discard <- qc.high_lib_size | qc.mito | qc.min_detected
DataFrame(LibSize=sum(qc.high_lib_size), Detected=sum(qc.min_detected),MitoProp=sum(qc.mito), Total=sum(discard))
```

---
## Removing low quality data

We will add this back in the colData slot to perform some final QC plots before filtering

```{r,fig.width=7,fig.height=4}
colData(sce.NeuroD1_filtered) <- cbind(colData(sce.NeuroD1_filtered),DataFrame(toDiscard=discard))
p1 <- plotColData(sce.NeuroD1_filtered,x="sum",y="detected",colour_by = "toDiscard")
p2 <- plotColData(sce.NeuroD1_filtered,x="sum",y="subsets_Mito_percent",colour_by = "toDiscard")
p3 <- plotColData(sce.NeuroD1_filtered,x="detected",y="subsets_Mito_percent",colour_by = "toDiscard")
gridExtra::grid.arrange(p1,p2,p3,ncol=3)
```

---
## Removing low quality data
At this point we may want to remove the low quality cells from further analysis

```{r,fig.width=7,fig.height=4}
sce.NeuroD1_filtered_QCed <- sce.NeuroD1_filtered[,sce.NeuroD1_filtered$toDiscard %in% "FALSE"]
sce.NeuroD1_filtered_QCed
```


---
# Log Normalisation

As with our RNA-seq and genomics data we want to normalise for total library size and perfrom a log2 transformation.

```{r,fig.width=7,fig.height=4}
sce.NeuroD1_filtered_QCed <- logNormCounts(sce.NeuroD1_filtered_QCed)
assayNames(sce.NeuroD1_filtered_QCed)
```


---
## Log Normalisation 2

An alternative approach for scRNA is to normalise accounting for potentially different cell-populations.

As we dont know cell populations yet, a quick clustering of data followed by this normalisation using clusters is performed.

```{r,fig.width=7,fig.height=4}
library(scran)
clust.sce.NeuroD1_filtered_QCed <- quickCluster(sce.NeuroD1_filtered_QCed) 
sce.NeuroD1_filtered_trimmed <- computeSumFactors(sce.NeuroD1_filtered_QCed, cluster=clust.sce.NeuroD1_filtered_QCed)
sce.NeuroD1_filtered_QCed <- logNormCounts(sce.NeuroD1_filtered_QCed)
assayNames(sce.NeuroD1_filtered_QCed)

```

---
## Modeling the mean-variance relationship

As with RNA-seq we want to account for mean-dispersion relationship. 

In Bioconductor we can use the *modelGeneVar* function to perform this for us.

```{r,fig.width=7,fig.height=4}
##
dec.NeuroD1_filtered_QCed <- modelGeneVar(sce.NeuroD1_filtered_QCed)
ggplot(dec.NeuroD1_filtered_QCed,aes(x=mean,y=total))+geom_point()
```

---
## Identifying top variable genes for PCA

Having fit the mean-dispersion relationship we can extract the most variable genes for use in PCA dimension reduction.

```{r,fig.width=7,fig.height=4}
top.NeuroD1_filtered_QCed <- getTopHVGs(dec.NeuroD1_filtered_QCed, n=3000)

```

---
## PCA

We can now perform a PCA on our data. 

The PCs will be used for clustering and 2D-representation of our data.
```{r,fig.width=7,fig.height=4}
set.seed(100) # See below.
sce.NeuroD1_filtered_QCed <- fixedPCA(sce.NeuroD1_filtered_QCed,
                                      subset.row=top.NeuroD1_filtered_QCed) 
reducedDimNames(sce.NeuroD1_filtered_QCed)

```

---
## PCA

We can produce a quick plot of the PCs to see if there are any biases immediately evident in the PCs.

```{r,fig.width=7,fig.height=4}
plotReducedDim(sce.NeuroD1_filtered_QCed, dimred="PCA",colour_by = "subsets_Mito_percent")

```

---
## Select PCs from PCA

For further analysis we want to select the PCs which explain most of the data.

To do this we can plot the variance explained by each PC and identify an elbow point.
```{r,fig.width=7,fig.height=4}
percent.var <- attr(reducedDim(sce.NeuroD1_filtered_QCed), "percentVar")
plot(percent.var, log="y", xlab="PC", ylab="Variance explained (%)")
# library(PCAtools)
# chosen.elbow <- findElbowPoint(percent.var)
```


---
## Visualise in 2D with Tisne and UMAP

With our PCs now calculated we can project these onto a 2D representation of the data. Here we will use two popular methods of TSNE and UMAP.

```{r,fig.width=7,fig.height=4}

sce.NeuroD1_filtered_QCed <- runTSNE(sce.NeuroD1_filtered_QCed,n_dimred=30)
sce.NeuroD1_filtered_QCed <- runUMAP(sce.NeuroD1_filtered_QCed,n_dimred=30)
reducedDimNames(sce.NeuroD1_filtered_QCed)
```

---
## Visualise in 2D with Tisne and UMAP

Witht the TSNE and UMAP we may first want to overlay some features of QC to see if this affects the data still.

```{r,fig.width=7,fig.height=4}
plotUMAP(sce.NeuroD1_filtered_QCed,colour_by="subsets_Mito_percent")
plotTSNE(sce.NeuroD1_filtered_QCed,colour_by="subsets_Mito_percent")
```

---
## Clustering

For clustering we can use the set of tools available in the [bluster](https://www.bioconductor.org/packages/release/bioc/html/bluster.html) package.

Here we run the default graph based methods walktrap and louvain from the igraph package.

```{r,fig.width=7,fig.height=4}
require(bluster)
clust.louvain <- clusterCells(sce.NeuroD1_filtered_QCed, use.dimred="PCA", 
                                BLUSPARAM=NNGraphParam(cluster.fun="louvain",cluster.args = list(resolution=0.8)))
clust.default <- clusterCells(sce.NeuroD1_filtered_QCed, use.dimred="PCA")

```




---
## Clustering

Once we have the clustering we will add them to the colData for the object and replot the UMAP

```{r,fig.width=7,fig.height=4}
colLabels(sce.NeuroD1_filtered_QCed) <- clust.louvain
colData(sce.NeuroD1_filtered_QCed)$DefaultLabel <- clust.default
plotUMAP(sce.NeuroD1_filtered_QCed,colour_by="label")

```

---
## Clustering 

We can now plot the mito percent against clusters to see if any clusters are driven by the mito percent of droplets.

```{r,fig.width=7,fig.height=4}
plotColData(sce.NeuroD1_filtered_QCed,x="label",y="subsets_Mito_percent",colour_by = "label")
```
---

## Clustering

We can then compare our overlaps between our clusterings in a simple heatmap.

```{r,fig.width=7,fig.height=4}
tab <- table(Walktrap=clust.default, Louvain=clust.louvain)
rownames(tab) <- paste("Walktrap", rownames(tab))
colnames(tab) <- paste("louvain", colnames(tab))
library(pheatmap)
pheatmap(log10(tab+10), color=viridis::viridis(100), cluster_cols=FALSE, cluster_rows=FALSE)
```

---
## Some ground truth

The paper also provides their annotation of cells. We can read this in and overlay with our umap.

```{r,eval=TRUE}
eec_paper_meta <- read.delim("https://rubioinformatics.s3.amazonaws.com/scRNA_graduate/GSE224223_EEC_metadata.csv",sep=";")
nrd1_cells <- as.data.frame(eec_paper_meta[eec_paper_meta$Library == "Nd1",])
   
nrd1_cells$X <- gsub("_2$","",nrd1_cells$X)

newMeta <- merge(colData(sce.NeuroD1_filtered_QCed),nrd1_cells,by.x="Barcode",by.y="X",all.x=TRUE,all.y=FALSE,order=FALSE)
rownames(newMeta) <- newMeta$Barcode
newMeta$All_Cell_Types[is.na(newMeta$All_Cell_Types)] <- "Filtered"
colData(sce.NeuroD1_filtered_QCed) <- newMeta
plotUMAP(sce.NeuroD1_filtered_QCed,colour_by="All_Cell_Types")

```

---
## Some ground truth

We can also see how the annotation overlaps with our clusters.

```{r,fig.width=7,fig.height=4}
tab <- table(Walktrap=sce.NeuroD1_filtered_QCed$All_Cell_Types, Louvain=sce.NeuroD1_filtered_QCed$label)
rownames(tab) <- paste("CellType", rownames(tab))
colnames(tab) <- paste("Leiden", colnames(tab))
library(pheatmap)
pheatmap(log10(tab+10), color=viridis::viridis(100), cluster_cols=FALSE, cluster_rows=FALSE)
```

---
## Identify markers

Now we have clusters defined we can identify genes which are differentially expressed within a cluster compared to other clusters.

```{r,fig.width=7,fig.height=4}
markers.sce.NeuroD1 <- scoreMarkers(sce.NeuroD1_filtered_QCed,sce.NeuroD1_filtered_QCed$label)
chosen <- markers.sce.NeuroD1[["4"]]
ordered <- chosen[order(chosen$mean.AUC, decreasing=TRUE),]
head(ordered[,1:4]) # showing basic stats only, for brevity.
```

---
## Plot markers

We can then use the *plotExpression* function to visualise a gene's expression across clusters.

```{r,fig.width=7,fig.height=4}
plotExpression(sce.NeuroD1_filtered_QCed, features=head(rownames(ordered),n=1), 
    x="label", colour_by="label")
```

```{r}
saveRDS(sce.NeuroD1_filtered_QCed,file="sce.NeuroD1_filtered_QCed.RDS")
```

---

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Seurat

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Seurat

---
"    
  )
  
}

```


---

## Working with CellRanger filtered matrix

We can use Seurat to import our data as before for Bioconductor.

```{r,fig.width=7,fig.height=4}
library(Seurat)
h5file <- "https://rubioinformatics.s3.amazonaws.com/scRNA_graduate/SRR_NeuroD1/outs/filtered_feature_bc_matrix.h5"
local_h5file <- basename(h5file)
download.file(h5file,local_h5file)
Nd1T.mat <- Read10X_h5(filename = local_h5file)
Nd1T.obj <- CreateSeuratObject(Nd1T.mat, project = "Nd1")

```
  
  
---  

## The Seurat object

The Seurat object contains now our information on Neurod1 droplets.


```{r,fig.width=7,fig.height=4}
Nd1T.obj
```
    
   
--- 

## The Seurat object

We can access rownames and column names in a similar manner as before.

```{r,fig.width=7,fig.height=4}
rownames(Nd1T.obj)[1:2]
colnames(Nd1T.obj)[1:2]
```
    
---   
## The Seurat object

To access metadata we need to directly access the slot

```{r,fig.width=7,fig.height=4}
Nd1T.obj@meta.data[1:2,]
```
    
---   
## The Seurat object

We can also review the active assay and active cell-barcode identifiers.

```{r,fig.width=7,fig.height=4}
Nd1T.obj@active.assay
Nd1T.obj@active.ident[1:2]
```
 
---   
## The Seurat object

The get access to the assays we can pull directly from the relevent slots.

```{r,fig.width=7,fig.height=4}
Nd1T.obj@assays$RNA$counts[1:2,1:2]
```
    
 
---   
## Gather QC

As with Biocondcutor we must again identify mitochondrial genes. 

```{r,fig.width=7,fig.height=4}
mito.genes <- grep("^MT", rownames(Nd1T.obj), value = T)
mito.genes[1:10]
```

---   
## Gather QC

We can the use *PercentageFeatureSet* function to identify percent of mito reads in each cell and the *AddMetaData* function to add this to the Seurat object.

```{r,fig.width=7,fig.height=4}
percent.mt <- PercentageFeatureSet(Nd1T.obj, features = mito.genes)
Nd1T.obj <- AddMetaData(Nd1T.obj,metadata = percent.mt,col.name = "percent.mt")
head(Nd1T.obj@meta.data)
```

---   
## Plot QC

We can then review QC metrics across our droplets.

```{r,fig.width=7,fig.height=4}
VlnPlot(Nd1T.obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

---   
## Plot QC

As well as plot QC metrics against each other.

```{r,fig.width=7,fig.height=4}
plot1 <- FeatureScatter(Nd1T.obj, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Nd1T.obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3 <- FeatureScatter(Nd1T.obj, feature1 = "nFeature_RNA", feature2 = "percent.mt")
plot1+plot2+plot3
```
---
## Subset by QC

We use the subset function to remove low quality cells from our object.

```{r,fig.width=7,fig.height=4}

Nd1T.obj.filt <- subset(Nd1T.obj, subset = `nCount_RNA` < 125000 & percent.mt < 25 & nFeature_RNA > 200) 
Nd1T.obj.filt
```
---
## Normalise data

We will now normalise our data by scaling to library size and applying a log2 transformation.

```{r,fig.width=7,fig.height=4}
Nd1T.obj.filt <- NormalizeData(Nd1T.obj.filt)
```
---
## Variable features

As with Bioc we will now identify variable features. 

Here the mean-variance relationship is accounted for by applying a VST transformation.

```{r,fig.width=7,fig.height=4}
Nd1T.obj.filt <- FindVariableFeatures(Nd1T.obj.filt, selection.method = "vst", nfeatures = 3000)
top10 <- head(VariableFeatures(Nd1T.obj.filt), 10)
top10
```
---
## Variable features

We can then plot these features using the **VariableFeaturePlot** function.

```{r,fig.width=7,fig.height=4}
plot1 <- VariableFeaturePlot(Nd1T.obj.filt)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```
---
## Scaling data

In Bioconductor the scaling of data was done inside the PCA functions.

Here we apply the scaling using the **ScaleData** function.

```{r,fig.width=7,fig.height=4}
all.genes <- rownames(Nd1T.obj.filt)
Nd1T.obj.filt <- ScaleData(Nd1T.obj.filt, features = all.genes)
Nd1T.obj.filt@assays
```
---
## PCA 

Now we have our scaled data we can apply PCA and identify how many PCs to use in downstream analysis.

```{r,fig.width=7,fig.height=4}
Nd1T.obj.filt <- RunPCA(Nd1T.obj.filt, features = VariableFeatures(object = Nd1T.obj))
ElbowPlot(Nd1T.obj.filt,ndims = 50)
```
---
## UMAP 

We can then create a 2D representation of our data using a UMAP .


```{r,fig.width=7,fig.height=4}
Nd1T.obj.filt <- RunUMAP(Nd1T.obj.filt, dims = 1:30)
```
---
## UMAP 

And we plot this using the *FeaturePlot* function overlaying the percent.mt.

```{r,fig.width=7,fig.height=4}
FeaturePlot(Nd1T.obj.filt, reduction = "umap",features = "percent.mt")
```
---
## Clustering 

We can then perform a louvain clustering using the *FindNeighbours* and *FindClusters* functions.

The cluster labels will be stored in active.idents slots.

```{r,fig.width=7,fig.height=4}
Nd1T.obj.filt <- FindNeighbors(Nd1T.obj.filt, dims = 1:30)
Nd1T.obj.filt <- FindClusters(Nd1T.obj.filt, resolution = 0.7)
Nd1T.obj.filt@active.ident
```
---
## Clustering 

We then visualise the clustering using the Dimplot fucntion which will use the active.ident slot by default.


```{r,fig.width=7,fig.height=4}
DimPlot(Nd1T.obj.filt, reduction = "umap")
```
---
# Compare clustering

We can quickly compare the clusters from Bioconductor and Seurat.

```{r,fig.width=7,fig.height=4}
bioc.clusters <- data.frame(Bioc=sce.NeuroD1_filtered_QCed$label,row.names = sce.NeuroD1_filtered_QCed$Barcode)
seurat.clusters <- as.data.frame(Nd1T.obj.filt@active.ident)
colnames(seurat.clusters) <- "Seurat"
clustersAll <- merge(bioc.clusters,seurat.clusters,by=0,all=FALSE)

```
---
# Compare clustering

```{r,fig.width=7,fig.height=4}
tab <- table(Bioc=clustersAll$Bioc, Seurat=clustersAll$Seurat)
rownames(tab) <- paste("Bioc", rownames(tab))
colnames(tab) <- paste("Seurat", colnames(tab))
library(pheatmap)
pheatmap(log10(tab+10), color=viridis::viridis(100), cluster_cols=FALSE, cluster_rows=FALSE)

```
---
# Find Markers

The *FindAllMarkers* function allows us to identify markers across all clusters.

```{r,fig.width=7,fig.height=4}
Nd1T.markers <- FindAllMarkers(Nd1T.obj.filt, only.pos = TRUE)
head(Nd1T.markers[Nd1T.markers$cluster == 1,])
```

---
# Find Markers

And we can then visualise this using the VlnPlot function.

```{r,fig.width=7,fig.height=4}
VlnPlot(Nd1T.obj.filt, features = "Pzp")

```
 
 