---
title: "Exercise 1"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/scRNA-seq"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
suppressPackageStartupMessages(require(Seurat))
```


<p>&nbsp;</p>
These exercises are about manipulate single-cell data with Seurat. Please download the counting matrix from [BOX](https://rockefeller.box.com/s/d6dvkjljprvqyglul1zxudv0ktfvwhsg) and loading them into a Seurat object. Or you may also used the rds file *data/scSeq_CKO_1kCell_ori.rds*. 

```{r loading_data,include=FALSE,eval=FALSE}
library(Seurat)
tar_dir <- "~/Desktop/01_scSeq_testCode_Bioconductor/CKO"
samID <- "CKO"
X10_file <- Seurat::Read10X(tar_dir)
obj <- CreateSeuratObject(counts = X10_file, project = samID, min.cells = 5,min.features = 200)
obj$dset <- samID
obj <- Seurat::RenameCells(obj,add.cell.id=samID)
cellID <- sample(rownames(obj@meta.data),size = 1000,replace = FALSE)
obj <- obj[,cellID]
saveRDS(obj,"~/Documents/packDev_ALL/scRNA-seq/scRNASeq/inst/extdata/data/scSeq_CKO_1kCell_ori.rds")
```


**Exercise 1** - **Data manipulation with Seurat**

## loading data
1. Please create a Seurat object by reading counting data from BOX or by loading the rds file. Then calculate mitochondrial contents of each ell
```{r,echo=toMessage}
library(Seurat)
obj <- readRDS("data/scSeq_CKO_1kCell_ori.rds")
obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^mt-")
head(obj@meta.data)
```

## Basi QC plot
2. Please access the read counts (nCount_RNA), gene counts (nFeature_RNA), and mitochondrial content (percent.mt) for each cell and draw violin plot
```{r,echo=toMessage}
VlnPlot(obj,features = c("nCount_RNA","nFeature_RNA","percent.mt"),pt.size = 0.2)
```

3. Please made a dot plot for nCount_RNA vs nFeature_RNA and try to point out potential doublets
```{r,echo=toMessage}
FeatureScatter(obj,feature1="nCount_RNA",feature2="nFeature_RNA")
```

4. Please made a dot plot for nount_RNA vs percent.mt and try to point out potential cell debris
```{r,echo=toMessage}
FeatureScatter(obj,feature1="nCount_RNA",feature2="percent.mt")
```

5. Please remove cells with percent.mt >= 10 for following analysis
```{r,echo=toMessage}
mt_cutH <- 10
obj_unfiltered <- obj
obj <- subset(obj,subset = percent.mt < mt_cutH)
obj
```

## Cell cycle
6. Please estimate cell cycle phase of each cell and a table to describe how many cells per phase
```{r,echo=toMessage}
convertHumanGeneList <- function(x){
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  genesV2 = getLDS(attributes = c("hgnc_symbol"), 
                   filters = "hgnc_symbol", 
                   values = x , 
                   mart = human,
                   attributesL = c("mgi_symbol"), 
                   martL = mouse, uniqueRows=T)
  humanx <- unique(genesV2[, 2])
  return(humanx)}
#
s_gene <- convertHumanGeneList(cc.genes$s.genes)
g2m_gene <- convertHumanGeneList(cc.genes$g2m.genes)
#
obj <- CellCycleScoring(obj, s.features = s_gene,
                              g2m.features = g2m_gene, set.ident = TRUE)
table(obj@meta.data$Phase)
```

## Normalization and clustering
7. Please scale data regressing to mitochondrial content (percent.mt) and cell cycle (S.score, G2M.score, Phase).
```{r,echo=toMessage}
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj,vars.to.regress = c("percent.mt","S.score","G2M.score","Phase"))
```

8. Please make priciple component analysis (PCA), estimate how many PCs would best represent this data, then make clustering and UMAP plot
- checkin 30 PCs
- elbow plot can be used to determine PCs
```{r,echo=toMessage}
set.seed(0)
obj <- RunPCA(obj, npcs = 30, verbose = FALSE)
ElbowPlot(obj,ndims = 30)
#
numPC <- 10
maxPC <- numPC
obj <- FindNeighbors(obj, reduction = "pca", dims = 1:maxPC)
obj <- FindClusters(obj, resolution = 0.5)
#
obj <- RunUMAP(obj, reduction = "pca", dims = 1:maxPC)
obj <- RunTSNE(obj, reduction = "pca", dims = 1:maxPC)
#
DimPlot(obj,reduction = "umap")
```

## Marker genes
9. Try to select marker genes for each cluster and generate heatmap of top five marker genes
```{r,echo=toMessage}
obj <- SetIdent(obj,value = "seurat_clusters")
clust.markers <- FindAllMarkers(obj, 
                                only.pos = TRUE,
                                min.pct = 0.25, 
                                logfc.threshold = 0.25)
head(clust.markers)
#
topG <- clust.markers %>% 
  group_by(cluster) %>% 
  top_n(n = 5, wt = avg_log2FC)
#
DoHeatmap(obj, features = topG$gene) + NoLegend()
```

## Advanced plots
10. Please display the top marker gene of each cluster in FeaturePlot and RidgePlot
```{r,echo=toMessage}
topG <- clust.markers %>% 
  group_by(cluster) %>% 
  top_n(n = 1, wt = avg_log2FC)
top_gene <- topG$gene
#
FeaturePlot(obj,features = top_gene,pt.size = 0.2)
RidgePlot(obj,features = top_gene)
```
