<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Session1.knit</title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <script src="libs/header-attrs-2.10/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Single-cell RNA sequencing ~ Session 1
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/scRNA-seq/" class="uri">http://rockefelleruniversity.github.io/scRNA-seq/</a>

---




## Overview
In this course we are going to introduce basic analysis for single-cell RNAseq, with a specific focus on the *10X system*. The course is divided into three sessions. In the first session, we will introduce how to interpret the **Cell  Ranger QC report** and how to do analysis with the **LOUPE Browser**. We will also demonstrate the customized analysis we can support right know. In the Second session, we will demonstrate how to process scRNAseq data and make QC reports with **Seurat**. In the third session, we will discuss how to to do more advanced analysis and QC with **Bioconductor** packages. 

---
class: inverse, center, middle

# Cell Ranger results

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Cell Ranger
[Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) is a widely used software for single-cell sequencing data analysis. It supports various kinds of analysis for *10X single-cell sequencing data*.
- Alignment and counting
- Analysis of VDJ changes
- Datasets aggregation
- Convert BCL files (raw files) into FASTQ files
- Generation of customized GTF and Genome files

---
## Basic reports of Cell Ranger
- [QC Report](../../data/web_summary.html)
- cloupe file: can be processed by [LOUPE Browser](https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser)

&lt;img src="imgs/cell_count.png" width="50%" style="display: block; margin: auto;" /&gt;

---
## Session information
&lt;img src="imgs/session_info.png" width="75%" style="display: block; margin: auto;" /&gt;

---
## Sequencing QC
&lt;img src="imgs/seq_qc.png" width="75%" style="display: block; margin: auto;" /&gt;

---
## Mapping QC
&lt;img src="imgs/map_info.png" width="75%" style="display: block; margin: auto;" /&gt;

---
## General information
&lt;img src="imgs/cell_count.png" width="75%" style="display: block; margin: auto;" /&gt;

- How many cells should we get?
- Is the throughput enough?

*It depends on the sample characteristics and library complexities*

---
## How to determin cell numbers?
&lt;img src="imgs/knee_plot.png" width="50%" style="display: block; margin: auto;" /&gt;

---
## Knee plot
- [Knee plot](https://liorpachter.wordpress.com/tag/knee-plot/) is applied to determine real cell numbers in a single-cell cohort.
- The x-axis represented the barcodes **ranked by UMI counts** inside the cell barcodes. Y-axis represents the UMI counts detected per cell barcode.
- A knee plot usually contains two bumps. We would take the **mid-point** of the first bump as a cut-off to differentiate real cells and backgrounds (empty droplets).
- As we defined the real cells and empty droplets, we could estimate the transcripts detected in real cell or empty droplets. 
- The transcripts in empty droplet, so call **ambient RNA**, are cell-free RNAs falsely included in droplets. We will demonstrate how to remove interference of ambient RNAs  in session 3.

---
## How to determine throughput?
- Downsampling and evaluate median gene per cells in different throughput (Mean reads per cell)
- Downsampling and evaluate proportion of unique UMIs in different throughput (Mean reads per cell)

---
## genes vs throughput
&lt;img src="imgs/gene_eval.png" width="75%" style="display: block; margin: auto;" /&gt;

*Seems not saturated yet*

---
## Saturation vs throughput
- Saturation: uniquely detected UMIs in total UMI
- Can evaluate both saturation and library complexity, like PCR duplication

&lt;img src="imgs/umi_eval.png" width="50%" style="display: block; margin: auto;" /&gt;

*Seems still not saturated and with lower library complexity, probably higher PCR duplicates.*

---
## tSNE plot and clustering
&lt;img src="imgs/tsne.png" width="100%" style="display: block; margin: auto;" /&gt;

*Seems the cells were grouped by UMI counts.*

---
## Data visualization with LOUPE Browser
- Please install the latest version
- How to access the expression of particular genes
- How to use split view
- How to define sub-groups
- How to calculate differential gene expression
- Could I import projections and more information generated by external programs?

---
class: inverse, center, middle

# Customized Analysis

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Demonstration
Here we use an article published by Fuchs Lab as an example to demonstrate the customized analysis we have done here at the BRC. Please access this [article](https://elifesciences.org/articles/56980). 
- General QC plots
- Data integration and normalization
- Clustering
- Annotate cell types
- Pseudotime analysis
- Differential gene expression and single-cell pathway analysis

---
## The next two sessions
We will also use the data of the eLife paper in next two sessions. You can download the data from GEO [link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE147415) or from DropBox [link](https://www.dropbox.com/sh/ztu0pucvu21szxm/AABGUhTryKp1T1CsoEUFRPcwa?dl=0).
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
