<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Session4.knit</title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <script src="libs/header-attrs-2.10/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Single-cell RNA sequencing ~ Session 3
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="https://rockefelleruniversity.github.io/scRNA-seq/" class="uri">https://rockefelleruniversity.github.io/scRNA-seq/</a>

---




## Overview

---
class: inverse, center, middle

# Loading data and empty droplets

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Install python libraries

```r
if (!require(reticulate)) {
    install.packages("reticulate")
    require(reticulate)
}
```

```
## Loading required package: reticulate
```

```r
if (!dir.exists(reticulate::miniconda_path())) {
    install_miniconda()
}
```

```
## * Downloading 'https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh' ...
```

```
## * Installing Miniconda -- please wait a moment ...
```

```
## * Miniconda has been successfully installed at '/github/home/.local/share/r-miniconda'.
```

```
## [1] "/github/home/.local/share/r-miniconda"
```

```r
conda_install(packages = c("python-igraph", "scanpy", "louvain"))
```

---
## Use conda python

```r
use_condaenv()
py_config()
```

```
## python:         /github/home/.local/share/r-miniconda/envs/r-reticulate/bin/python
## libpython:      /github/home/.local/share/r-miniconda/envs/r-reticulate/lib/libpython3.6m.so
## pythonhome:     /github/home/.local/share/r-miniconda/envs/r-reticulate:/github/home/.local/share/r-miniconda/envs/r-reticulate
## version:        3.6.13 | packaged by conda-forge | (default, Feb 19 2021, 05:36:01)  [GCC 9.3.0]
## numpy:          /github/home/.local/share/r-miniconda/envs/r-reticulate/lib/python3.6/site-packages/numpy
## numpy_version:  1.19.5
```

---
## Load python library

```r
sc &lt;- import("scanpy")
sc
```

```
## Module(scanpy)
```

---
## Install anndata package

```r
if (!require(anndata)) {
    install.packages("anndata")
    require(anndata)
}
```

```
## Loading required package: anndata
```

```
## Warning in library(package, lib.loc = lib.loc, character.only = TRUE,
## logical.return = TRUE, : there is no package called 'anndata'
```

```
## Installing package into '/usr/local/lib/R/host-site-library'
## (as 'lib' is unspecified)
## Loading required package: anndata
```

---
## Download PBMC

```r
pathToDownload &lt;- "http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz"
download.file(pathToDownload, "pbmc3k_filtered_gene_bc_matrices.tar.gz")
utils::untar("pbmc3k_filtered_gene_bc_matrices.tar.gz")
```

---
## Read PBMC

```r
adata &lt;- sc$read_10x_mtx("filtered_gene_bc_matrices/hg19/", var_names = "gene_ids")

adata
```

```
## AnnData object with n_obs × n_vars = 2700 × 32738
##     var: 'gene_symbols'
```

---
## Dimensions PBMC

```r
dim(adata)
```

```
## [1]  2700 32738
```

---
## Dimnames - PBMC

```r
colnames(adata)[1:3]
```

```
## [1] "ENSG00000243485" "ENSG00000237613" "ENSG00000186092"
```

```r
rownames(adata)[1:3]
```

```
## [1] "AAACATACAACCAC-1" "AAACATTGAGCTAC-1" "AAACATTGATCAGC-1"
```

---
## Counts- PBMC

```r
countsMatrix &lt;- adata$X
countsMatrix[1:2, 1:2]
```

```
## 2 x 2 sparse Matrix of class "dgTMatrix"
##                  ENSG00000243485 ENSG00000237613
## AAACATACAACCAC-1               .               .
## AAACATTGAGCTAC-1               .               .
```

---
## Gene information - PBMC

```r
var_df &lt;- adata$var
var_df[1:2, , drop = FALSE]
```

```
##                 gene_symbols
## ENSG00000243485   MIR1302-10
## ENSG00000237613      FAM138A
```

---
## Cell information - PBMC

```r
obs_df &lt;- adata$obs
obs_df
```

```
## data frame with 0 columns and 2700 rows
```

```r
rownames(obs_df)[1:2]
```

```
## [1] "AAACATACAACCAC-1" "AAACATTGAGCTAC-1"
```

---
## Cell information - PBMC

```r
sc$pp$filter_cells(adata, min_genes = 0)
sc$pp$filter_genes(adata, min_cells = 0)
```

---
## Cell information - PBMC

```r
var_df &lt;- adata$var
var_df[1:2, ]
```

```
##                 gene_symbols n_cells
## ENSG00000243485   MIR1302-10       0
## ENSG00000237613      FAM138A       0
```

---
## Cell information - PBMC

```r
obs_df &lt;- adata$obs
obs_df[1:2, , drop = FALSE]
```

```
##                  n_genes
## AAACATACAACCAC-1     781
## AAACATTGAGCTAC-1    1352
```

---
## Cell information - PBMC

```r
var_df$mito &lt;- grepl("MT-", var_df$gene_symbols)
var_df[var_df$mito == TRUE, ]
```

```
##                 gene_symbols n_cells mito
## ENSG00000254959 INMT-FAM188B       0 TRUE
## ENSG00000198888       MT-ND1    2558 TRUE
## ENSG00000198763       MT-ND2    2416 TRUE
## ENSG00000198804       MT-CO1    2686 TRUE
## ENSG00000198712       MT-CO2    2460 TRUE
## ENSG00000228253      MT-ATP8      32 TRUE
## ENSG00000198899      MT-ATP6    2014 TRUE
## ENSG00000198938       MT-CO3    2647 TRUE
## ENSG00000198840       MT-ND3     557 TRUE
## ENSG00000212907      MT-ND4L     398 TRUE
## ENSG00000198886       MT-ND4    2588 TRUE
## ENSG00000198786       MT-ND5    1399 TRUE
## ENSG00000198695       MT-ND6     249 TRUE
## ENSG00000198727       MT-CYB    2517 TRUE
```
---

## Cell information - PBMC

```r
adata$var &lt;- var_df
adata$var[1:2, ]
```

```
##                 gene_symbols n_cells  mito
## ENSG00000243485   MIR1302-10       0 FALSE
## ENSG00000237613      FAM138A       0 FALSE
```

---

## Cell information - PBMC

```r
sc$pp$calculate_qc_metrics(adata,
                           qc_vars = list("mito"),
                           inplace = TRUE)
adata
```

```
## AnnData object with n_obs × n_vars = 2700 × 32738
##     obs: 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito'
##     var: 'gene_symbols', 'n_cells', 'mito', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
```

---

## Cell information - PBMC

```r
adata$var[5:10,]
```

```
##                  gene_symbols n_cells  mito n_cells_by_counts mean_counts
## ENSG00000239945  RP11-34P13.8       0 FALSE                 0 0.000000000
## ENSG00000237683    AL627309.1       9 FALSE                 9 0.003333333
## ENSG00000239906 RP11-34P13.14       0 FALSE                 0 0.000000000
## ENSG00000241599  RP11-34P13.9       0 FALSE                 0 0.000000000
## ENSG00000228463    AP006222.2       3 FALSE                 3 0.001111111
## ENSG00000237094 RP4-669L17.10       0 FALSE                 0 0.000000000
##                 log1p_mean_counts pct_dropout_by_counts total_counts
## ENSG00000239945       0.000000000             100.00000            0
## ENSG00000237683       0.003327790              99.66667            9
## ENSG00000239906       0.000000000             100.00000            0
## ENSG00000241599       0.000000000             100.00000            0
## ENSG00000228463       0.001110494              99.88889            3
## ENSG00000237094       0.000000000             100.00000            0
##                 log1p_total_counts
## ENSG00000239945           0.000000
## ENSG00000237683           2.302585
## ENSG00000239906           0.000000
## ENSG00000241599           0.000000
## ENSG00000228463           1.386294
## ENSG00000237094           0.000000
```

---

## Cell information - PBMC

```r
adata$obs[1:2,]
```

```
##                  n_genes n_genes_by_counts log1p_n_genes_by_counts total_counts
## AAACATACAACCAC-1     781               781                6.661855         2421
## AAACATTGAGCTAC-1    1352              1352                7.210080         4903
##                  log1p_total_counts pct_counts_in_top_50_genes
## AAACATACAACCAC-1           7.792349                   47.74886
## AAACATTGAGCTAC-1           8.497807                   45.50275
##                  pct_counts_in_top_100_genes pct_counts_in_top_200_genes
## AAACATACAACCAC-1                    63.27964                    74.96902
## AAACATTGAGCTAC-1                    61.02386                    71.81318
##                  pct_counts_in_top_500_genes total_counts_mito
## AAACATACAACCAC-1                    88.39323                73
## AAACATTGAGCTAC-1                    82.62288               186
##                  log1p_total_counts_mito pct_counts_mito
## AAACATACAACCAC-1                4.304065        3.015283
## AAACATTGAGCTAC-1                5.231109        3.793596
```


---
## Cell information - PBMC

```r
sc$pl$highest_expr_genes(adata, gene_symbols = "gene_symbols")
```


---
## Time for an exercise!

Exercise on scRNAseq analysis with Bioconductor can be found [here](../../exercises/exercises/exercise4_exercise.html).

Answers can be found [here](../../exercises/answers/exercise4_answers.html).
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
