<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Session3.utf8</title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Single-cell RNA sequencing ~ Session 3
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="https://rockefelleruniversity.github.io/scRNA-seq/" class="uri">https://rockefelleruniversity.github.io/scRNA-seq/</a>

---




## Outline
There are more and more Bioconductor packages supporting single-cell data analysis. R Amezquita, A Lun, S Hicks, and R Gottardo wrote an integrated workflow, [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/release/OSCA/), for single-cell data analysis and quality assessment. In this session, we will go through several important QC metrics which can't be made with Seurat. 

- How to differentiate empty droplets?
- How to estimate ambient RNA and remove them?
- How to identify doublets?
- Checking for confounded effect?

---
class: inverse, center, middle

# Loading data and empty droplets

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Load data from Cell Ranger Result
We can load the data in from 10X with the *DropletUtils* function, *read10xCounts()*.

```r
library(DropletUtils)
library(DropletTestFiles)
fname &lt;- "~/filepath/toCellRanger/results"
sce &lt;- read10xCounts(fname, col.names = TRUE)
cellID &lt;- colData(sce)$Barcode
```

---
## Subset dataset
We have a small subset version of this dataset that you can load in from the *data/* directory to try this out. 



---
## SingleCellExperiment Object

*read10xCounts()* reads the data in as a specialist object called a SingleCellExperiment. 

```r
sce
```

```
## class: SingleCellExperiment 
## dim: 27998 100000 
## metadata(1): Samples
## assays(1): counts
## rownames(27998): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000096730 ENSMUSG00000095742
## rowData names(2): ID Symbol
## colnames(100000): CGTGTCTTCGCATGAT-1 TTTATGCGTCGAACAG-1 ...
##   TTATGCTGTGTTTGTG-1 GCTGCGACACGTTGGC-1
## colData names(2): Sample Barcode
## reducedDimNames(0):
## altExpNames(0):
```

---
## SingleCellExperiment Object
The *colData()* and *rowData()* functions can be used to access experiment metadata. 


```r
# cell information
colData(sce)[1:2, ]
```

```
## DataFrame with 2 rows and 2 columns
##                                    Sample            Barcode
##                               &lt;character&gt;        &lt;character&gt;
## CGTGTCTTCGCATGAT-1 ~/Desktop/01_scSeq_t.. CGTGTCTTCGCATGAT-1
## TTTATGCGTCGAACAG-1 ~/Desktop/01_scSeq_t.. TTTATGCGTCGAACAG-1
```

```r
# gene information
rowData(sce)[1:2, ]
```

```
## DataFrame with 2 rows and 2 columns
##                                    ID      Symbol
##                           &lt;character&gt; &lt;character&gt;
## ENSMUSG00000051951 ENSMUSG00000051951        Xkr4
## ENSMUSG00000089699 ENSMUSG00000089699      Gm1992
```

---
## Access UMI counts in each droplet

```r
bcrank &lt;- barcodeRanks(counts(sce))
bcrank[1:2, ]
```

```
## DataFrame with 2 rows and 3 columns
##                         rank     total    fitted
##                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## CGTGTCTTCGCATGAT-1   41200.5         1        NA
## TTTATGCGTCGAACAG-1   41200.5         1        NA
```

---
## Knee plot

```r
uniq &lt;- !duplicated(bcrank$rank)
# 
plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", 
    cex.lab = 1.2)

abline(h = metadata(bcrank)$inflection, col = "darkgreen", lty = 2)
abline(h = metadata(bcrank)$knee, col = "dodgerblue", lty = 2)

legend("bottomleft", legend = c("Inflection", "Knee"), col = c("darkgreen", "dodgerblue"), 
    lty = 2, cex = 1.2)
```

---
## Knee Plot

```
## Warning in xy.coords(x, y, xlabel, ylabel, log): 1 y value &lt;= 0 omitted from
## logarithmic plot
```

![](Session3_files/figure-html/rankUMI_knee2-1.png)&lt;!-- --&gt;

---
## Identify non-empty droplets
  The *emptydrops* function can be used to identify goog/bad droplets. 2 keys arguments are:
- limit: lowest counts per cell
- test.ambient: Could be used to estimate ambient RNA contamination


```r
set.seed(100)
limit &lt;- 100
e.out &lt;- emptyDrops(counts(sce), lower = limit, test.ambient = TRUE)
# 
e.out
```

```
## DataFrame with 100000 rows and 5 columns
##                        Total     LogProb     PValue   Limited        FDR
##                    &lt;integer&gt;   &lt;numeric&gt;  &lt;numeric&gt; &lt;logical&gt;  &lt;numeric&gt;
## CGTGTCTTCGCATGAT-1         1    -5.34170 0.70912909     FALSE 0.95697174
## TTTATGCGTCGAACAG-1         1    -6.44503 0.50414959     FALSE 0.92603663
## CATGACACAAGTCTGT-1         0          NA         NA        NA         NA
## ACATACGCACCACCAG-1        58  -266.20222 0.01839816     FALSE 0.45640979
## CGTCCATAGCTGCAAG-1      1601 -2917.18710 0.00009999      TRUE 0.00507393
## ...                      ...         ...        ...       ...        ...
## GTGCTTCGTCACCCAG-1         0          NA         NA        NA         NA
## GGAAAGCCAAGGCTCC-1         1    -9.06371 0.20007999     FALSE   0.818381
## TGCCCTAGTAGCTCCG-1         1   -13.51757 0.00309969     FALSE   0.125927
## TTATGCTGTGTTTGTG-1         0          NA         NA        NA         NA
## GCTGCGACACGTTGGC-1         0          NA         NA        NA         NA
```

---
## Identify non-empty droplets

```r
# Testeed by FDR
summary(e.out$FDR &lt;= 0.001)
```

```
##    Mode   FALSE    TRUE    NA's 
## logical   54010     591   45399
```

```r
# Concordance by testing with FDR and limited
table(Sig = e.out$FDR &lt;= 0.001, Limited = e.out$Limited)
```

```
##        Limited
## Sig     FALSE  TRUE
##   FALSE 53525   485
##   TRUE      1   590
```

---
## Distribution of non-empty reads
We can plot the distribution of significance of non-empty reads


```r
hist(e.out$PValue[e.out$Total &lt;= limit &amp; e.out$Total &gt; 0], xlab = "P-value", main = "", 
    col = "grey80")
```

![](Session3_files/figure-html/estDroplet3-1.png)&lt;!-- --&gt;

---
## Subset non-empty droplets
We can filter to just our non-empty droplets using a simple which query on the FDR from *emptyDrops()*. Here we are using a 0.001 cut-off. 

```r
sce2 &lt;- sce[, which(e.out$FDR &lt;= 0.001)]
```


---
class: inverse, center, middle

# Normalization and clustering

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Counts normalization

```r
library(scran)
library(scuttle)
library(scater)
clusters &lt;- quickCluster(sce2)
sce2 &lt;- computeSumFactors(sce2, cluster = clusters)
sce2 &lt;- logNormCounts(sce2)
sce2
```

```
## class: SingleCellExperiment 
## dim: 27998 591 
## metadata(1): Samples
## assays(2): counts logcounts
## rownames(27998): ENSMUSG00000051951 ENSMUSG00000089699 ...
##   ENSMUSG00000096730 ENSMUSG00000095742
## rowData names(2): ID Symbol
## colnames(591): GCTCCTAAGGCACATG-1 GTAACTGTCGGATGGA-1 ...
##   CACCAGGTCACCTTAT-1 AGGTCATCATGTAAGA-1
## colData names(3): Sample Barcode sizeFactor
## reducedDimNames(0):
## altExpNames(0):
```

---
## Identify variable features

```r
set.seed(1000)
# modeling variables
dec.pbmc &lt;- modelGeneVarByPoisson(sce2)
# calcualte top features
top.pbmc &lt;- getTopHVGs(dec.pbmc, prop = 0.1)
```

---
## Make TSNE and UMAP plots

```r
set.seed(1000)
# Evaluate PCs
sce2 &lt;- denoisePCA(sce2, subset.row = top.pbmc, technical = dec.pbmc)
# make TSNE plot
sce2 &lt;- runTSNE(sce2, dimred = "PCA")
# make UMAP plot
sce2 &lt;- runUMAP(sce2, dimred = "PCA")
```

---
## Graphic based clustering

```r
g &lt;- buildSNNGraph(sce2, k = 10, use.dimred = "PCA")
clust &lt;- igraph::cluster_walktrap(g)$membership
colLabels(sce2) &lt;- factor(clust)
# 
colData(sce2)
```

```
## DataFrame with 591 rows and 4 columns
##                                    Sample            Barcode sizeFactor
##                               &lt;character&gt;        &lt;character&gt;  &lt;numeric&gt;
## GCTCCTAAGGCACATG-1 ~/Desktop/01_scSeq_t.. GCTCCTAAGGCACATG-1   0.913482
## GTAACTGTCGGATGGA-1 ~/Desktop/01_scSeq_t.. GTAACTGTCGGATGGA-1   0.970723
## CGGACACCATTACGAC-1 ~/Desktop/01_scSeq_t.. CGGACACCATTACGAC-1   0.725295
## GCGAGAAAGCTACCGC-1 ~/Desktop/01_scSeq_t.. GCGAGAAAGCTACCGC-1   0.700158
## AAACGGGCAGATGGGT-1 ~/Desktop/01_scSeq_t.. AAACGGGCAGATGGGT-1   1.636871
## ...                                   ...                ...        ...
## GTGGGTCAGCACCGTC-1 ~/Desktop/01_scSeq_t.. GTGGGTCAGCACCGTC-1   1.026750
## CGCTATCAGTACGACG-1 ~/Desktop/01_scSeq_t.. CGCTATCAGTACGACG-1   0.677916
## GTACTTTTCCTTTCGG-1 ~/Desktop/01_scSeq_t.. GTACTTTTCCTTTCGG-1   0.978623
## CACCAGGTCACCTTAT-1 ~/Desktop/01_scSeq_t.. CACCAGGTCACCTTAT-1   1.130759
## AGGTCATCATGTAAGA-1 ~/Desktop/01_scSeq_t.. AGGTCATCATGTAAGA-1   1.689680
##                       label
##                    &lt;factor&gt;
## GCTCCTAAGGCACATG-1        2
## GTAACTGTCGGATGGA-1        2
## CGGACACCATTACGAC-1        2
## GCGAGAAAGCTACCGC-1        6
## AAACGGGCAGATGGGT-1        4
## ...                     ...
## GTGGGTCAGCACCGTC-1        6
## CGCTATCAGTACGACG-1        6
## GTACTTTTCCTTTCGG-1        1
## CACCAGGTCACCTTAT-1        7
## AGGTCATCATGTAAGA-1        4
```

---
## UMAP plot

```r
plotUMAP(sce2, colour_by = "label")
```

![](Session3_files/figure-html/plotUMAP_2-1.png)&lt;!-- --&gt;

---
## tSNE plot

```r
plotTSNE(sce2, colour_by = "label")
```

![](Session3_files/figure-html/plotTSNE-1.png)&lt;!-- --&gt;

---
class: inverse, center, middle

# Removing Ambient RNA

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Ambient RNA
- Cell-free RNAs can contaminate droplets.
- They can be estimated by empty droplets.

```r
# extrat potential ambient RNA and thee estimated score
amb &lt;- metadata(e.out)$ambient[, 1]
head(amb)
```

```
## ENSMUSG00000051951 ENSMUSG00000025902 ENSMUSG00000033845 ENSMUSG00000025903 
##       5.808442e-07       5.808442e-07       1.067188e-04       6.783426e-05 
## ENSMUSG00000033813 ENSMUSG00000002459 
##       5.876210e-05       5.808442e-07
```

---
## Remove ambient RNA

```r
library(scater)
stripped &lt;- sce2[names(amb), ]
out &lt;- removeAmbience(counts(stripped), ambient = amb, groups = colLabels(stripped))
```

---
## Integrate corrected counts

```r
counts(stripped, withDimnames = FALSE) &lt;- out
stripped &lt;- logNormCounts(stripped)
```

---
## Before/After removal
- Hemoglobin A1 (Hba-a1) as example
- In most cases the Hbs are contaminated from residual RBCs

```r
ensmbl_id &lt;- rowData(sce2)$ID[rowData(sce2)$Symbol == "Hba-a1"]
plotExpression(sce2, x = "label", colour_by = "label", features = ensmbl_id) + ggtitle("Before")

plotExpression(stripped, x = "label", colour_by = "label", features = ensmbl_id) + 
    ggtitle("After")
```

---
## Before/After removal
Hemoglobin A1 (Hba-a1) as example

.pull-left[
![](Session3_files/figure-html/compRMAmb_pres1-1.png)&lt;!-- --&gt;
  ]

.pull-right[
![](Session3_files/figure-html/compRMAmb_pres2-1.png)&lt;!-- --&gt;
  ]
  
---
## Before/After removal
- Krt17 as example

```r
ensmbl_id &lt;- rowData(sce2)$ID[rowData(sce2)$Symbol == "Krt17"]
plotExpression(sce2, x = "label", colour_by = "label", features = ensmbl_id) + ggtitle("Before")
plotExpression(stripped, x = "label", colour_by = "label", features = ensmbl_id) + 
    ggtitle("After")
```

---
## Before/After removal
Krt17 as example

.pull-left[
![](Session3_files/figure-html/compRMAmb2_pres-1.png)&lt;!-- --&gt;
  ]

.pull-right[
![](Session3_files/figure-html/compRMAmb2_pres2-1.png)&lt;!-- --&gt;
  ]

---
## Save result
Again we can save our results for later using a Rds object.

```r
saveRDS(stripped, "scSeq_CTRL_sceSub_rmAmbRNA.rds")
```


---
class: inverse, center, middle

# Removing Doublets

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## Doublets
- Doublets means two or more cells clumped in a single droplet. Thus, the read counts and genes detected in this droplet would be much higher than other droplets.
- Here, we demonstrate how to identify doublets using simulation with [scran](https://bioconductor.org/packages/release/bioc/manuals/scran/man/scran.pdf)

---
## Normalization and clustering

```r
dec &lt;- modelGeneVar(stripped)
hvgs &lt;- getTopHVGs(dec, n = 1000)
stripped &lt;- runPCA(stripped, ncomponents = 10, subset_row = hvgs)
stripped &lt;- runUMAP(stripped, dimred = "PCA")
g &lt;- buildSNNGraph(stripped, k = 10, use.dimred = "PCA")
clust &lt;- igraph::cluster_walktrap(g)$membership
colLabels(stripped) &lt;- factor(clust)
plotUMAP(stripped, colour_by = "label")
```

---
## Normalization and clustering
![](Session3_files/figure-html/normClust2-1.png)&lt;!-- --&gt;

---
## Estimate doublets
The *computeDoubletDensity()* fucntion can be used on your SingleCellExperiment object to estimate doublets. 

```r
dbl.dens &lt;- computeDoubletDensity(stripped, #subset.row=top.mam, 
    d=ncol(reducedDim(stripped)),subset.row=hvgs)
summary(dbl.dens)
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.2908  0.7978  1.1442  1.2442  1.6010  4.1961
```

```r
stripped$DoubletScore &lt;- dbl.dens
```

---
## Plot doublets scores ~ UMAP
We can project the doublet scores onto our UMPA to see if doublet correlates with the data distribution.

```r
plotUMAP(stripped, colour_by = "DoubletScore")
```

![](Session3_files/figure-html/doublerScore-1.png)&lt;!-- --&gt;

---
## Plot doublets scores by cluster
Likewise we can look for over representation of doublets in clusters.

```r
plotColData(stripped, x = "label", y = "DoubletScore", colour_by = "label") + geom_hline(yintercept = quantile(colData(stripped)$DoubletScore, 
    0.95), lty = "dashed", color = "red")
```

![](Session3_files/figure-html/doublerScorebyClust-1.png)&lt;!-- --&gt;
- No clusters have significantly higher doublet scores than other clusters. No clusters would be removed.
- Red dash line represented 95% quantile of doublet score. The cells with higher doublet score than this cut-off would be removed.

---
## Remove doublets
We can clean our data up based on this 95% quantile cut-off. 

```r
cut_off &lt;- quantile(stripped$DoubletScore, 0.95)
stripped$isDoublet &lt;- c("no", "yes")[factor(as.integer(stripped$DoubletScore &gt;= cut_off), 
    levels = c(0, 1))]
table(stripped$isDoublet)
```

```
## 
##  no yes 
## 561  30
```

```r
sce_clean &lt;- stripped[stripped$isDoublet == "no", ]
saveRDS(sce_clean, "scSeq_CTRL_sceSub_rmAmbRNA_rmDoublet.rds")
```



---
class: inverse, center, middle

# QC plots after clearance

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## QC plots
- Mitochondrial content
- Genes detected
- Reads per cell

```r
library(scater)
mtGene &lt;- rowData(sce_clean)$ID[grepl(rowData(sce_clean)$Symbol, pattern = "mt-")]
is.mito &lt;- names(sce_clean) %in% mtGene
sce_clean &lt;- addPerCellQC(sce_clean, subsets = list(Mito = is.mito))
```

---
## QC plots - Read counts


Post-clean

```r
plotColData(sce_clean, x = "label", y = "sum", colour_by = "label") + ggtitle("read counts")
```

![](Session3_files/figure-html/qc_mrgr-1.png)&lt;!-- --&gt;


---
## QC plots - Gene counts

Post-clean

```r
plotColData(sce_clean, x = "label", y = "detected", colour_by = "label") + ggtitle("gene counts")
```

![](Session3_files/figure-html/qc_mrg2r-1.png)&lt;!-- --&gt;


---
## QC plots - Mitochondrial %

Post-clean

```r
plotColData(sce_clean, x = "label", y = "subsets_Mito_percent", colour_by = "label") + 
    ggtitle("mitocondrial content")
```

![](Session3_files/figure-html/qc_mrg3r-1.png)&lt;!-- --&gt;


---
## QC plots ~ comparison
- Mitochondrial contents vs read counts

```r
plotColData(sce_clean, x = "sum", y = "subsets_Mito_percent", colour_by = "label") + 
    ggtitle("is.mito vs read counts")
```

![](Session3_files/figure-html/qc_complex-1.png)&lt;!-- --&gt;

---
## QC plots ~ comparison
- Gene counts vs read counts

```r
plotColData(sce_clean, x = "sum", y = "detected", colour_by = "label") + ggtitle("gene counts vs read counts")
```

![](Session3_files/figure-html/qc_complex2-1.png)&lt;!-- --&gt;

---
## Estimate variance explaination
- Clustering (label), mitochondrial content (subsets_Mito_percent), doublets (DoubletScore), read counts (sum), and gene counts (detected) were tested.
- We would suppose "label" (clustering) would explain more variances than other controls.

```r
vars &lt;- getVarianceExplained(sce_clean, variables = c("DoubletScore", "label", "sum", 
    "detected", "subsets_Mito_percent"))
```

```r
plotExplanatoryVariables(vars)
```


---
## Estimate variance explaination


```r
plotExplanatoryVariables(vars)
```

```
## Warning: Removed 2765 rows containing non-finite values (stat_density).
```

![](Session3_files/figure-html/varExp3-1.png)&lt;!-- --&gt;

---
## Exercise Time
- [Exercise]()
- [Answer]()
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
